splunk/packages/edit-table/src/components/KVStoreUploader.jsx

```diff
@@ -82,19 +92,71 @@ export default function KVStoreUploader({
             return false;
         }
 
-        const isCorrectFormat = checkJsonArrayCorrectFormat(jsonData, dataFields);
+        // Sometime, user may upload CSV file with _key field (because they don't want the _key that is auto-generated by Splunk)
+        // So, we need to remove _key field from dataFields if isIncludeKeyInCSV is disabled
+        const dataFieldsWithoutKey = dataFields.filter((field) => field !== '_key');
+        const actualDataFields = isIncludeKeyInCSV ? dataFields : dataFieldsWithoutKey;
+
+        const isCorrectFormat = checkJsonArrayCorrectFormat(jsonData, actualDataFields);
         if (!isCorrectFormat) {
             showErrorMessage(dataFieldErrorMsg);
             return false;
         }
 
-        const formattedJsonData = projectFields(jsonData, dataFields);
+        const formattedJsonData = projectFields(jsonData, actualDataFields);
         return formattedJsonData;
     };
 
-    const runUpload = async () => {
-        const data = await transformData();
-        if (!data) return;
+    // Build a map of primary key and entry. Eg. ERR-01 -> { code: ERR-01, name: Error 1, ... }
+    const buildEntriesMap = (entries) => {
+        const entriesMap = new Map();
+        entries.forEach((entry) => {
+            entriesMap.set(entry[primaryKey], entry);
+        });
+        return entriesMap;
+    };
+
+    const getIncrementalUploadData = async (newData) => {
+        // If primary key is not set, insert all new data
+        if (!primaryKey) {
+            return {
+                dataForUpdate: [],
+                dataForInsert: newData,
+            };
+        }
+
+        // If there is no data in KV store, insert all new data
+        const entries = await getAllCollectionEntries(splunkApp, collectionName, fetchAllErrorMsg);
+        if (entries == null || entries.length === 0) {
+            return {
+                dataForUpdate: [],
+                dataForInsert: newData,
+            };
+        }
```

Comment:
do we really need this? the if entries is empty, the build code below would exactly create
```
return {
                dataForUpdate: [],
                dataForInsert: newData,
            };
```

Reply:
then we should leave a comment for this - it's hard to understand :)